src/pocketmine/Player.php

```diff
@@ -2788,10 +2790,35 @@ public function handlePlayerAction(PlayerActionPacket $packet) : bool{
 				//TODO: destroy-progress level event
 				break;
 			case PlayerActionPacket::ACTION_START_SWIMMING:
-				break; //TODO
+				$block = $this->level->getBlock($this);
+
+				$ev = new PlayerToggleSwimEvent($this, true);
+				if(!($block instanceof Water)){
+					$ev->setCancelled();
+				}
+				$this->server->getPluginManager()->callEvent($ev);
+				if($ev->isCancelled()){
+					$this->sendData($this);
+				}else{
+					$this->propertyManager->setFloat(self::DATA_BOUNDING_BOX_HEIGHT, $this->width);
```

Comment:
this will produce astonishing behaviour if a plugin modifies the player's scale (this data property is supposed to be the base height, not rescaled).